core/src/main/java/com/linecorp/armeria/server/Http2RequestDecoder.java

```diff
@@ -182,6 +182,10 @@ public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers
                 ctx.fireChannelRead(req);
             }
         } else {
+            if (!(req instanceof DecodedHttpRequestWriter)) {
+                throw connectionError(PROTOCOL_ERROR,
```

Comment:
Question) Is it correct behavior to close the connection rather than return a 4xx response?

If a malicious user decides to continuously send these type of requests, will other requests be penalized?

Reply:
Good point. 
4xx response might be returned already by `HttpServerHandler`. So we don't need to return a response here. 
Is it better to silently ignore the incoming DATA frames?

Reply:
> If a malicious user decides to continuously send these type of requests, will other requests on the same connection be penalized?

For the protocol violations or unexpected errors, a connection is reset with `writeErrorResponse()`.
For example, if a malicious user sends a wrong `content-length`, the connection which the request was sent will be closed as well.
However, the `HttpServerHandler` who handled the request did not close the connection. So it makes more sense to keep the connection.

Reply:
> For example, if a malicious user sends a wrong content-length, the connection which the request was sent will be closed as well.

Maybe I misunderstood the current codebase, but wouldn't only the stream be closed in this case?

https://github.com/line/armeria/blob/b9dc1ad1c6f4cfee8aba8e50a61d203c37eb94cc/core/src/main/java/com/linecorp/armeria/server/Http2RequestDecoder.java#L150

Since the connection will be closed only if `streamId == 0`

![Screen Shot 2022-09-01 at 6 44 46 PM](https://user-images.githubusercontent.com/8510579/187884707-2c60d32a-804e-491e-8c94-29072e63469b.png)

Anyways, sorry about the late check. I agree with the approach 👍